apisix/init.lua

```diff
@@ -868,6 +869,70 @@ local function healthcheck_passive(api_ctx)
 end
 
 
+function _M.status()
+    core.response.exit(200, "ok")
+end
+
+function _M.status_ready()
+    local_conf = core.config.local_conf()
+    local provider = core.table.try_read_attr(local_conf, "deployment", "role_traditional",
+                                              "config_provider")
+    if provider == "yaml" then
+        local status_shdict = ngx.shared.status_report_standalone
+        local pids = status_shdict:get_keys()
+        local resp = ""
+
+        for _, pid in pairs(pids) do
+            local ready = status_shdict:get(pid)
+            if not ready then
+                core.log.warn("worker pid: ", pid,
+                    " has not received configuration")
+                resp = resp .. "Worker pid: " .. pid .. " has not received configuration\n"
+                break
+            end
+        end
+
+        if resp and resp ~= "" then
+            core.response.exit(503, resp)
+            return
+        end
+
+        core.response.exit(200, "ok")
+        return
+    end
+    if provider == "etcd" then
+        local status_shdict = ngx.shared.status_report
+        local pids = status_shdict:get_keys()
+        local resp = ""
+        for _, pid in pairs(pids) do
+            local need_reload = status_shdict:get(pid)
+            if need_reload then
+                resp = resp .. "Worker pid: " .. pid .. " is out of sync with etcd\n"
+            end
+        end
+
+        if resp and resp ~= "" then
+            core.response.exit(503, resp)
+            return
+        end
+
+        local yaml_conf = local_conf
+        local res, err
+        for _, host in ipairs(yaml_conf.etcd.host) do
+            res, err = etcd_util.request(host .. "/version", yaml_conf)
+            if res then
+                core.response.exit(200)
+                return
+            end
+        end
```

Comment:
What does etcd reachability have to do with whether an instance is READY?

This is a complete departure from the APISIX design intent of using old configurations to provide services when etcd is unavailable to ensure that services are not interrupted. 

If we use it in Kubernetes Probes, as soon as etcd becomes unavailable, that APISIX instance is reported as unhealthy and thus stops receiving incoming traffic. If there are no replaceable instances, the service stops completely.

Reply:
This status ready API should be limited to whether the instance has loaded any configuration to make it work, either etcd or standalone. etcd availability or whether the configuration is up-to-date is not part of this scope.

Reply:
I agree with @bzp2010's opinion that the `/status/ready` endpoint in etcd mode should indeed not be related to the success or failure of connecting to etcd. In that case, it seems no conditions are needed in etcd mode, and a direct return of 200 is sufficient.